src/dcap_provider.cpp

```diff
@@ -714,3 +721,112 @@ extern "C" sgx_plat_error_t sgx_ql_set_logging_function(
     logger_callback = logger;
     return SGX_PLAT_ERROR_OK;
 }
+
+extern "C" sgx_plat_error_t sgx_get_qe_identity_info(
+    sgx_qe_identity_info_t** pp_qe_identity_info)
+{
+    std::string intel_cert_base_url = INTEL_QE_IDENTITY_URL;
+    sgx_qe_identity_info_t *p_qe_identity_info = NULL;
+    char* buffer = nullptr;
+
+    if (!pp_qe_identity_info)
+    {
+        log(SGX_QL_LOG_ERROR, "Invalid parameter pp_qe_identity_info"); 
+        return SGX_PLAT_ERROR_INVALID_PARAMETER;
+    }
+
+    try 
+    {
+        std::vector<uint8_t> identity_info;
+        std::string issuer_chain;
+        size_t total_buffer_size = 0;
+
+        const auto curl = curl_easy::create(intel_cert_base_url, CYBERTRUST_ROOT_CERT);
+        log(SGX_QL_LOG_INFO, "Fetching QE Identity from remote server: '%s'.", intel_cert_base_url.c_str());
+        curl->perform();
+
+        // content type
+        std::string content_type = curl->unescape(*curl->get_header(headers::CONTENT_TYPE));
+        log(SGX_QL_LOG_INFO, "CONTENT_TYPE:[%s]\n", content_type.c_str());
+
+        // issuer chain
+        const std::string* issuer_chain_raw_header = curl->get_header(headers::QE_ISSUER_CHAIN);
+        if (issuer_chain_raw_header == nullptr)
+        {
+            log(SGX_QL_LOG_ERROR,
+                "Header '%s' is missing.",
+                headers::QE_ISSUER_CHAIN);
+            return SGX_PLAT_ERROR_UNEXPECTED_SERVER_RESPONSE;
+        }
+        issuer_chain = curl->unescape(*issuer_chain_raw_header);
+        log(SGX_QL_LOG_INFO, "QE_ISSUER_CHAIN: [%s]\n", issuer_chain.c_str());
+
+        // requester id
+        std::string request_id = curl->unescape(*curl->get_header(headers::REQUEST_ID));

```

Comment:
Missing null check (like above).

I think there should be a helper function to get unescaped headers that can catch missing headers, log them, and return `SGX_PLAT_ERROR_UNEXPECTED_SERVER_RESPONSE`.

Reply:
Checking added

Reply:
Please add a helper function instead of duplicating code.

I know it's only ~10 loc, but any form of duplication is bad news.

Reply:
Okay! :-)
Added get_unescape_header() and get_raw_header().